File: backend/app/schemas/analytics.py
"""Analytics response schemas."""
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class EventsByHomeItem(BaseModel):
    """Event count by home."""

    home_id: str
    count: int


class DeviceUptimeItem(BaseModel):
    """Device uptime summary item."""

    device_id: str
    event_count: int
    last_event: Optional[str] = None
    uptime_percent: float


class OpsOverviewResponse(BaseModel):
    """Operations overview response."""

    totalHomes: int
    totalDevices: int
    totalDevicesOnline: int
    openAlertsBySeverity: dict[str, int]  # {"high": 0, "medium": 0, "low": 0}
    eventsByHomeLast24h: list[EventsByHomeItem]
    deviceUptimeSummary: list[DeviceUptimeItem]


class OwnerOverviewResponse(BaseModel):
    """Owner overview response."""

    openAlertsCount: int
    openAlertsHigh: int
    devicesOnlineCount: int
    eventsLast24h: int
    totalDevices: int
    roomsCount: Optional[int] = None
    perRoomStats: Optional[list[dict]] = None  # Stub for future use


class AlertsHeatmapItem(BaseModel):
    """Alerts heatmap item."""

    home_id: str
    home_name: str
    total_alerts: int
    alerts_by_severity: dict[str, int]  # {"high": 0, "medium": 0, "low": 0}


class AlertsHeatmapResponse(BaseModel):
    """Alerts heatmap response."""

    period: str  # "24h" or "7d"
    data: list[AlertsHeatmapItem]


class EventsByHourItem(BaseModel):
    """Events by hour item."""

    hour: int
    day: int
    month: int
    year: int
    count: int

File: frontend/src/api/analytics.ts
/** RTK Query endpoints for analytics. */
import { api } from "./base";
import { AlertsHeatmapResponse, OpsOverviewResponse, OwnerOverviewResponse } from "../types";

export const analyticsApi = api.injectEndpoints({
  endpoints: (builder) => ({
    getOwnerOverview: builder.query<OwnerOverviewResponse, { home_id: string }>({
      query: ({ home_id }) => ({
        url: `/owner/overview?home_id=${home_id}`,
        method: "GET",
      }),
      providesTags: ["Alert", "Device"],
    }),
    getOpsOverview: builder.query<OpsOverviewResponse, void>({
      query: () => ({
        url: "/ops/overview",
        method: "GET",
      }),
      providesTags: ["Alert", "Device", "Home"],
    }),
    getAlertsHeatmap: builder.query<AlertsHeatmapResponse, { period?: "24h" | "7d" }>({
      query: ({ period = "24h" }) => ({
        url: `/ops/alerts/heatmap?period=${period}`,
        method: "GET",
      }),
      providesTags: ["Alert"],
    }),
    getOwnerEventsTimeseries: builder.query<
      { data: Array<{ hour: number; day: number; month: number; year: number; count: number }> },
      { home_id: string; hours?: number }
    >({
      query: ({ home_id, hours = 24 }) => ({
        url: `/owner/events/timeseries?home_id=${home_id}&hours=${hours}`,
        method: "GET",
      }),
      providesTags: ["Alert"],
    }),
  }),
});

export const {
  useGetOwnerOverviewQuery,
  useGetOpsOverviewQuery,
  useGetAlertsHeatmapQuery,
  useGetOwnerEventsTimeseriesQuery,
} = analyticsApi;

